### PR TITLE
Global-workflow (AR) Generic updates for Gaea C5

### PR DESCRIPTION
**Pull request by AnilKumar-NOAA**
# Global-Workflow-AR

# Description 
- Port global-workflow’s build and run capability to Gaea-C5 
- Building global-workflow on Gaea-C5
- Setting up experiments with global-workflow on Gaea-C5
- CI testing on Gaea C5

# Type of change

- Setting up global workflow for AR project on Gaea C5
- Spack-Stack for Gaea C5 
- Maintenance (clean-up, new CI test, etc.)
- To build: update scripts in sorc/ufs_utils.fd, sorc/ufs_model.fd, and sorc/gfs_utils.fd 
- To set up experiment: update workflow/ directory

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# Build global-workflow:
```
source workflow/gw_setup.sh
```
While in /sorc folder:
```
./build_all.sh
```
While in /sorc folder:
```
./link_workflow.sh
```

# How has this been tested?

Example:
- Forecast-only, coupled  on Gaea C5
- update CI scripts to test in SCRON environment 

### Atm-only:
- cd workflow 
```
./setup_expt.py gfs forecast-only --resdetatmos 96 --idate 2021032312 --edate 2021032312 --gfs_cyc 4 --pslot scron_test --comroot /gpfs/f5/epic/scratch/Anil.Kumar/COMROOT --expdir /gpfs/f5/epic/scratch/Anil.Kumar/EXPDIR
```
#### [https://global-workflow.readthedocs.io/en/latest/setup.html#id3](https://global-workflow.readthedocs.io/en/latest/setup.html#id3)
Go to your EXPDIR and check/change the following variables within your config.base now before running the next script:
### Run workflow generator script
```
./setup_xml.py $EXPDIR/$PSLOT
```
For example
```
./setup_xml.py /gpfs/f5/epic/scratch/Anil.Kumar/EXPDIR/scron_test --maxtries 1
```
## Start your run
```
rocotorun -d $PSLOT.db -w $PSLOT.xml
for example
rocotorun -w scron_test1.xml -d scron_test1.db
```

### Set up your experiment scron
```
Like the Gaea documentation suggests, you must add SCRON statements to your scrontab 

#SCRON --partition=cron_c5
#SCRON --job-name=<give a job name> 
#SCRON --chdir=/ncrc/proj/epic/rocoto/1.3.6/bin/rocotorun
#SCRON --output=/ncrc/home1/Anil.Kumar/scron_test1.%j.scrontab
*/5 * * * * /ncrc/proj/epic/rocoto/1.3.6/bin/rocotorun

The last line should be the time interval and the location of the script you want to run.

```
### scrontab file uses scron statement (for example)
```
scrontab -e 
```
####  modify/edit the file  
```
#SCRON --partition=cron_c5
#SCRON --time=1:00:00
#SCRON --mem=8GB
#SCRON --account=epic
#SCRON --job-name=scron_test
#SCRON --chdir=./
#SCRON --output=scron_test.%j.scrontab
#################### scron_test ####################
MAILTO=""
*/5 * * * * /ncrc/proj/epic/rocoto/1.3.6/bin/rocotorun -d /gpfs/f5/epic/scratch/Anil.Kumar/EXPDIR/scron_test/scron_test.db -w /gpfs/f5/epic/scratch/Anil.Kumar/EXPDIR/scron_test2/scron_test.xml
#################################################################
```
### - under workflow dir
```
./rocoto_viewer.py -w /gpfs/f5/epic/scratch/Anil.Kumar/EXPDIR/scron_test/scron_test.xml -d /gpfs/f5/epic/scratch/Anil.Kumar/EXPDIR/scron_test/scron_test.db

```

### CODE REVIEW

File: ush/detect_machine.sh [hunk 1]
```diff
@@ -26,6 +26,8 @@ case $(hostname -f) in
   gaea9.ncrc.gov)      MACHINE_ID=gaea ;; ### gaea9
   gaea1[0-6].ncrc.gov) MACHINE_ID=gaea ;; ### gaea10-16
 
+  gaea5[1-8].ncrc.gov) MACHINE_ID=gaea ;; ### gaea51-58
+
```
Comment (aerorahul) [inline ush/detect_machine.sh:30]:
```suggestion
gaea5[1-8].ncrc.gov) MACHINE_ID=gaea ;; ### gaea51-58

```
I think it is ok to remove the lines 24-27 for old Gaea

File: ush/detect_machine.sh [hunk 2]
```diff
@@ -26,6 +26,8 @@ case $(hostname -f) in
   gaea9.ncrc.gov)      MACHINE_ID=gaea ;; ### gaea9
   gaea1[0-6].ncrc.gov) MACHINE_ID=gaea ;; ### gaea10-16
 
+  gaea5[1-8].ncrc.gov) MACHINE_ID=gaea ;; ### gaea51-58
```
Comment (aerorahul) [inline ush/detect_machine.sh:25]:
```suggestion
  gaea5[1-8])          MACHINE_ID=gaea ;; ### gaea51-58
  gaea5[1-8].ncrc.gov) MACHINE_ID=gaea ;; ### gaea51-58
```
To stay in sync w/ the ufs-weather-model.
It is ok to remove lines 24-27 about the old Gaea

File: versions/fix.ver [hunk 3]
```diff
@@ -18,6 +18,6 @@ export mom6_ver=20231219
 export orog_ver=20231027
 export reg2grb2_ver=20220805
 export sfc_climo_ver=20220805
-export ugwd_ver=20231027
+export ugwd_ver=20220805
```
Comment (aerorahul) [inline versions/fix.ver]: This needs to be reverted to the `develop` version as it is not Gaea related.

File: parm/config/gfs/config.base [hunk 4]
```diff
@@ -70,8 +70,8 @@ export DO_BUFRSND="NO"     # BUFR sounding products
 export DO_GEMPAK="NO"      # GEMPAK products
 export DO_AWIPS="NO"       # AWIPS products
 export DO_NPOESS="NO"      # NPOESS products
-export DO_TRACKER="YES"    # Hurricane track verification
-export DO_GENESIS="YES"    # Cyclone genesis verification
+export DO_TRACKER="NO"    # Hurricane track verification
+export DO_GENESIS="NO"    # Cyclone genesis verification
```
Comment (aerorahul) [inline parm/config/gfs/config.base]: Please revert and add instructions to disable for Gaea.

File: env/GAEA.env [hunk 5]
```diff
@@ -0,0 +1,308 @@
+#! /usr/bin/env bash
+
+if [[ $# -ne 1 ]]; then
+
+    echo "Must specify an input argument to set runtime environment variables!"
+    echo "argument can be any one of the following:"
+    echo "atmanlrun atmensanlrun aeroanlrun snowanl"
+    echo "anal sfcanl fcst post metp"
+    echo "eobs eupd ecen efcs epos"
+    echo "postsnd awips gempak"
```
Comment (aerorahul) [inline env/GAEA.env]:
```suggestion
    echo "argument can be any one of the following:"
```
This is incomplete list, so might as well remove all of them.

File: env/GAEA.env [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+# Configure MPI environment
+export MPI_BUFS_PER_PROC=2048
+export MPI_BUFS_PER_HOST=2048
+export MPI_GROUP_MAX=256
+export MPI_MEMMAP_OFF=1
+export MP_STDOUTMODE="ORDERED"
+export KMP_AFFINITY=scatter
+export OMP_STACKSIZE=2048000
+export NTHSTACK=1024000000
+#export LD_BIND_NOW=1
```
Comment (aerorahul) [inline env/GAEA.env]: Are these appropriate for Gaea?  
Seems to be copied from Hera which has a very different architecture.
What happens if you remove these and let the defaults set in?

File: env/GAEA.env [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+ulimit -a
+
+if [[ "${step}" = "prep" ]] || [[ "${step}" = "prepbufr" ]]; then
+
+    nth_max=$((npe_node_max / npe_node_prep))
+
+    export POE="NO"
+    export BACK=${BACK:-"YES"}
+    export sys_tp="GAEA"
+    export launcher_PREP="srun"
```
Comment (aerorahul) [inline env/GAEA.env]: Is obs-processing enabled on Gaea?  If not, please remove and add when the capability is enabled.

File: env/GAEA.env [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+
+elif [[ "${step}" = "prepsnowobs" ]]; then
+
+    export APRUN_CALCFIMS="${launcher} -n 1"
```
Comment (aerorahul) [inline env/GAEA.env]: Is JEDI DA enabled on Gaea?  If not, please remove and add when the capability is enabled.

File: env/GAEA.env [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+    [[ ${NTHREADS_OCNANAL} -gt ${nth_max} ]] && export NTHREADS_OCNANAL=${nth_max}
+    export APRUN_OCNANAL="${launcher} -n ${npe_ocnanalchkpt} --cpus-per-task=${NTHREADS_OCNANAL}"
+
+elif [[ "${step}" = "ocnanalecen" ]]; then
+
+    nth_max=$((npe_node_max / npe_node_ocnanalecen))
+
+    export NTHREADS_OCNANALECEN=${nth_ocnanalecen:-${nth_max}}
+    [[ ${NTHREADS_OCNANALECEN} -gt ${nth_max} ]] && export NTHREADS_OCNANALECEN=${nth_max}
+    export APRUN_OCNANALECEN="${launcher} -n ${npe_ocnanalecen} --cpus-per-task=${NTHREADS_OCNANALECEN}"
```
Comment (aerorahul) [inline env/GAEA.env]: Is JEDI DA enabled on Gaea?  If not, please remove and add when the capability is enabled.

File: env/GAEA.env [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+    if [[ ${CDUMP} == "gfs" ]]; then
+        npe_gempak=${npe_gempak_gfs}
+        npe_node_gempak=${npe_node_gempak_gfs}
+    fi
+
+    nth_max=$((npe_node_max / npe_node_gempak))
+
+    export NTHREADS_GEMPAK=${nth_gempak:-1}
+    [[ ${NTHREADS_GEMPAK} -gt ${nth_max} ]] && export NTHREADS_GEMPAK=${nth_max}
+    export APRUN="${launcher} -n ${npe_gempak} ${mpmd_opt}"
```
Comment (aerorahul) [inline env/GAEA.env]: Are these downstream jobs enabled on Gaea?  If not, please remove and add when capabilities get added/tested

File: env/GAEA.env [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+
+    export CFP_MP=${CFP_MP:-"YES"}
+    export USE_CFP=${USE_CFP:-"YES"}
+    export APRUNCFP="${launcher} -n \$ncmd ${mpmd_opt}"
+
+    nth_max=$((npe_node_max / npe_node_eupd))
+
+    export NTHREADS_ENKF=${nth_eupd:-${nth_max}}
+    [[ ${NTHREADS_ENKF} -gt ${nth_max} ]] && export NTHREADS_ENKF=${nth_max}
+    export APRUN_ENKF="${launcher} -n ${npe_enkf:-${npe_eupd}} --cpus-per-task=${NTHREADS_ENKF}"
```
Comment (aerorahul) [inline env/GAEA.env]: Is Gaea GSI/EnKF cycling ready?  If not, please remove and add when capabilities are enabled.

File: ci/platforms/config.gaea [hunk 12]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/bash
+
+export GFS_CI_ROOT=/gpfs/f5/epic/proj-shared/global/GFS_CI_ROOT
+export ICSDIR_ROOT=/gpfs/f5/epic/proj-shared/global/glopara/data/ICSDIR
+export STMP="/gpfs/f5/epic/scratch/${USER}"
+export SLURM_ACCOUNT=epic
+export max_concurrent_cases=5
+export max_concurrent_pr=4
```
Comment (aerorahul) [inline ci/platforms/config.gaea]: Is the global-workflow CI running on Gaea?  If not, please add this in the PR that turns the CI ON on Gaea.

Comment (DavidBurrows-NCO) [reply ci/platforms/config.gaea]: @aerorahul @AnilKumar-NOAA Anil, you can remove this file..I have a branch for CI on Gaea that will include this.

**Review by AnilKumar-NOAA — COMMENTED** [review]
We have made all suggested changes. Tested and running successfully. Please check. Thanks epic team

File: ush/module-setup.sh [hunk 13]
```diff
@@ -70,39 +70,10 @@ elif [[ ${MACHINE_ID} = stampede* ]] ; then
...
-        module use -a /opt/cray/pe/craype/default/modulefiles
-    fi
-    if [[ -s /etc/opt/cray/pe/admin-pe/site-config ]] ; then
-        source /etc/opt/cray/pe/admin-pe/site-config
-    fi
-    if [[ "${__ms_source_etc_profile}" == yes ]] ; then
-        source /etc/profile
-        unset __ms_source_etc_profile
+	source /usr/share/lmod/lmod/init/bash
+	source /etc/profile
```
Comment (aerorahul) [inline ush/module-setup.sh]: Please maintain indentation

File: modulefiles/module_base.gaea.lua [hunk 14]
```diff
@@ -0,0 +1,52 @@
+help([[
+Load environment to run GFS on Gaea
+]])
+
+local spack_mod_path=(os.getenv("spack_mod_path") or "None")
+local spack_gsi_mod_path=(os.getenv("spack_gsi_mod_path") or "None")
+prepend_path("MODULEPATH", spack_mod_path)
+prepend_path("MODULEPATH", spack_gsi_mod_path)
```
Comment (aerorahul) [inline modulefiles/module_base.gaea.lua]: The gsi_mod_path contains the other as well.  I don't think the previous is needed.  Please confirm.

File: versions/run.gaea.ver [hunk 15]
```diff
@@ -0,0 +1,14 @@
+export stack_intel_ver=2023.1.0
+export stack_cray_mpich_ver=8.1.25
+export spack_env=unified-env
+export spack_gsi_env=gsi-addon-dev
```
Comment (aerorahul) [inline versions/run.gaea.ver]: See above comment in `module_base.gaea.lua`.
Only one env should be loaded.

**Review by aerorahul — CHANGES_REQUESTED** [review]
Please see comments re. spack env.
Also, please revert the ugwd version from fix.ver.

**Review by AnilKumar-NOAA — COMMENTED** [review]
All suggested changes are updated in this PR. Please review it. Thanks

**Comment by DavidBurrows-NCO** [comment]
Hi @aerorahul . In order to submit the jobs to compute nodes on Gaea, we have been manually altering the xml file to include this line: `<native>--clusters=c5</native>`
I see these native commands are inserted into the xml via `workflow/rocoto/tasks.py` and `workflow/hosts/gaea.yaml`. I don't see a way to get an additional Gaea-specific native setting passed into the xml file. Do you have a suggestion for this problem? Thanks!

**Comment by aerorahul** [comment]
> All suggested changes are updated in this PR. Please review it. Thanks

The `ugwd` version hasn't been reverted.

**Comment by aerorahul** [comment]
> Hi @aerorahul . In order to submit the jobs to compute nodes on Gaea, we have been manually altering the xml file to include this line: `<native>--clusters=c5</native>` I see these native commands are inserted into the xml via `workflow/rocoto/tasks.py` and `workflow/hosts/gaea.yaml`. I don't see a way to get an additional Gaea-specific native setting passed into the xml file. Do you have a suggestion for this problem? Thanks!

@DavidBurrows-NCO 
I'll think about this.  We used the `partition` flag for things like these, but seems like Gaea Slurm is set up differently.
Would you be able to post a snippet of the jobcard scheduler portion, and the corresponding XML snippet?

**Review by AnilKumar-NOAA — COMMENTED** [review]
@aerorahul - please review the updated changes. Thanks

**Comment by DavidBurrows-NCO** [comment]
@aerorahul 
On Gaea, you submit jobs to the c5 compute cluster then the partition within c5 is batch.

Here is the sbatch portion:
```
#SBATCH --job-name=test_for_xmlsetup_gfsstage_ic_12
#SBATCH --account=ufs-ard
#SBATCH --qos=normal
#SBATCH --partition=batch
#SBATCH -t 00:15:00
#SBATCH --nodes=1-1
#SBATCH --tasks-per-node=1
#SBATCH --cpus-per-task=1
#SBATCH -o /gpfs/f5/epic/scratch/David.Burrows/comroot/test_for_xmlsetup/logs/2021032312/gfsstage_ic.log
#SBATCH --clusters=c5
#SBATCH --export=NONE
```
and corresponding xml snippet:
```
<jobname><cyclestr>test_for_xmlsetup_gfsstage_ic_@H</cyclestr></jobname>
<account>ufs-ard</account>
<queue>normal</queue>
<partition>batch</partition>
<walltime>00:15:00</walltime>
<nodes>1:ppn=1:tpp=1</nodes>
<native>--clusters=c5</native>
<native>--export=NONE</native>
```

**Review by aerorahul — APPROVED** [review]
looks good.

**Pull request merged**